clang/lib/AST/Expr.cpp

```diff
@@ -4852,6 +4862,10 @@ BinaryOperator::BinaryOperator(const ASTContext &Ctx, Expr *lhs, Expr *rhs,
   if (hasStoredFPFeatures())
     setStoredFPFeatures(FPFeatures);
   setDependence(computeDependence(this));
+  if (hasWrappingOperand(Ctx))
```

Comment:
Implementation-wise, it is OK for both attributes to exist on the same type.

`hasWrapsAttr()` cautiously checks for `no_wraps` before claiming we have `wraps`.
```c
bool QualType::hasWrapsAttr() const {
  return !isNull() && getTypePtr()->hasAttr(attr::Wraps) &&
         !getTypePtr()->hasAttr(attr::NoWraps);
}

```

But yeah you're right, this is confusing because if you look at an ast dump you'll see both attributes applied to the type and it may not be obvious which has precedence. I'll add a diagnostic and a sane default of keeping the `no_wraps` and dropping `wraps` for cases like this.

Reply:
As of https://github.com/llvm/llvm-project/pull/115094/commits/445f8c7c93484a5476d85e05a289e609e3f38a94, only one attribute will persist through to BinaryOperators. Thanks